vibravox/utils.py

```diff
@@ -116,3 +115,99 @@ def reverse(x):
+
+    Args:
+        speech_batch (List[torch.Tensor]): 
+            A list of clean speech samples. Each tensor should be 1-dimensional with shape `(time,)`.
+        noise_batch (List[torch.Tensor]): 
+            A list of noise samples. Each tensor should be 1-dimensional with shape `(time_noise,)`.
+            The length of each noise sample must be greater than or equal to the corresponding speech sample.
+        snr_range (List[float], optional): 
+            A list containing two floats representing the minimum and maximum SNR values in decibels (dB).
+            Defaults to `[-3.0, 5.0]`.
+
+    Returns:
+        Tuple[List[torch.Tensor], List[torch.Tensor]]:
+            - `corrupted_speech_batch`: 
+                A list of corrupted speech samples obtained by adding scaled noise to the clean speech.
+                Each tensor has the same shape as the corresponding input speech tensor `(time,)`.
+            - `noise_batch_scaled`: 
+                A list of scaled noise tensors used for mixing. Each tensor has the same shape as the 
+                corresponding input speech tensor `(time,)`.
+    """
+    # Input validation
+    if not isinstance(speech_batch, list) or not all(isinstance(t, torch.Tensor) for t in speech_batch):
+        raise TypeError("speech_batch must be a list of torch.Tensor")
+    if not isinstance(noise_batch, list) or not all(isinstance(t, torch.Tensor) for t in noise_batch):
+        raise TypeError("noise_batch must be a list of torch.Tensor")
+    if len(speech_batch) != len(noise_batch):
+        raise ValueError("speech_batch and noise_batch must have the same length")
+
+    corrupted_speech_batch: List[torch.Tensor] = []
+    noise_batch_scaled: List[torch.Tensor] = []
+    
+    a, b = snr_range
+
+    number_segments: int = 10
+
+    k = (b - a)/number_segments
+
+    for speech, noise in zip(speech_batch, noise_batch):
+        
+        # Compute power
+        speech_power = torch.mean(speech ** 2)
+        noise_power = torch.mean(noise ** 2)
+        
+        if speech.dim() != 1:
+            raise ValueError(f"Each speech sample must be a 1D tensor, but got shape {speech.shape}")
+        if noise.dim() != 1:
+            raise ValueError(f"Each noise sample must be a 1D tensor, but got shape {noise.shape}")
+
+        time_speech = speech.size(0)
+        time_noise = noise.size(0)
+
+        if time_noise < time_speech:
+            raise ValueError(f"noise_sample length ({time_noise}) must be >= speech_sample length ({time_speech})")
+               
+        # Shuffle 10 equisegments of noise
+        r = len(noise)%number_segments
+        
+        noise = noise[r:]
+        n = len(noise)
+        
+        noise = noise.view(number_segments, -1)
+        
+        permutation = torch.randperm(number_segments)
+        
+        # Sample SNRs
+        snrs = torch.tensor([a+i*k for i in range(number_segments)])[permutation]
```

Comment:
Over-engineered sampling + why are you sampling several SNRs in the for loop ?

```suggestion
        snrs = torch.empty(1).uniform_(snr_range[0], snr_range[1]).item()
```

Please try to simplify your function, (a,b,r,n,k,...)

( P.S. The suggestion is to draw into the for loop, even though it would be better to remove the for loop and mix the signal at the batched tensor level)

Reply:
Basic philosophy of mix_speech_and_noise is as follows :
- shuffle n equisegments of noise
- apply SNR to them
- speech = speech + noise_SNRed

What I do is (as code demonstrates) :
- view noise as 10 equisegments
- permute the segments
- compute then apply snr to the segments
- flatten the SNRed noise then take what's needed i.e. noise_SNRed[ : time_speech ]

The two first steps allow for high variability whereas I see the third step a bit overengineered i.e. why apply different SNRs to all segments when the first few segments are needed? because it is very easy to compute that way

Reply:
Why do you need to have 10 segments ?

Reply:
New mix_speech_and_noise code implements a simpler algorithm (as discussed) :
- Take a random slice in noise
- Apply random SNR to it
- speech = speech + noise_SNRed